### DISCUSSION TITLE
Is there any way to color argument names in function calls without disabling semantic token color?

### DISCUSSION THREAD

**Discussion by ZeroRin** [microsoft/pylance-release#3565] [Q&A]
![image](https://user-images.githubusercontent.com/14952735/199496406-827d2e3d-5d3e-49ef-9aef-6a86e7f2e507.png)
what I want to do is to color argument names in all function calls to a different color like line 5 in this picture. I managed to do it with the following setting:
```json
        "[Default Dark+]": {
            "textMateRules": [
                {
                    "name": "Parameter Names",
                    "scope": "variable.parameter.function-call.python",
                    "settings": {
                        "foreground": "#4FC1FF"
                    }
                }
            ]
        }
```
however, as shown in the picuture, this only affects args that does not exist in function definition. This is because they are not assigned with any semantic token type. For defined argument (i.e. bar), all its occurance (in function definition, body and call) are labeled as semantic token type "parameter" and are painted with the same color.

Is there a way to change its color in function call and keep everything else the same? maybe any trick to override semantic token color with textmate rules?

**Comment by rchiodo** [comment]
You can customize semantic token colors too. I did the following:

```json
  "editor.tokenColorCustomizations": {
    "[Default Light+]": {
      "textMateRules": [
        {
          "name": "Parameter Names",
          "scope": "variable.parameter.function-call.python",
          "settings": {
            "foreground": "#FF0000"
          }
        }
      ]
    }
  },
  "editor.semanticTokenColorCustomizations": {
    "[Default Light+]": {
      "rules": {
        "parameter": "#00ff26"
      }
    }
  }
```

Which generates these colors for me on Default Light+

![image](https://user-images.githubusercontent.com/19672699/199544795-224ffbc5-760c-402e-ad68-4e23838dfc06.png)

**Reply by rchiodo** [reply]
If you need to see what the semantic token is, you can use the `Developer : Inspect Editor Tokens and Scopes`

**Reply by ZeroRin** [reply]
The result is basically the same as I posted, with `bar` in the last line having a different color as `baa`, and the same color as `bar` in the function definition. What I want to do is to paint the last `bar` to red and keep other `bar`s untouched.  

These `bar`s are distinguishable with textmates, but they share the same semantic token, so I cannot paint them with different color.

**Reply by ZeroRin** [reply]
I want to change every token with `variable.parameter.function-call.python` to a unique color in order to see function calls like `foo(bar=bar)` more clearly. Assigning a color to `parameter` does not work because if I call `fooo(bar=bar)` in `foo`, both `bar`s will be considered as `parameter`

**Comment by ZeroRin** [comment]
I tried to create a better example to clarify my need, with the following code:
```python
def foo(bar):
    print(bar)

def fooo(bar):
    foo(bar=bar)

bar = 1
foo(bar=bar)

```
The following pictures are (from left to right):
1. The default behavior (with default dark+)
2. What I expect (I managed to do it by turning off semantic token color)
3. The best I can do without turning off semantic token color
![1667445581308_106630FF-4A9F-47ec-933C-D19344EC1E17](https://user-images.githubusercontent.com/14952735/199641749-13b59a96-bce9-4ede-bb22-2c40ee7151b7.png)![1667445497677_5DE62148-8432-49ab-B09A-45126964C400](https://user-images.githubusercontent.com/14952735/199641907-1bbb5f5a-9c43-41c6-a5b7-926783be40fe.png)![image](https://user-images.githubusercontent.com/14952735/199642276-5bb06121-7d87-4c64-951e-1b4c15967d88.png)

The trouble is about these two `bar`s:
![image](https://user-images.githubusercontent.com/14952735/199642367-87b3e371-ec9b-4eec-be95-aadac1a31b0f.png)![image](https://user-images.githubusercontent.com/14952735/199642415-0f750195-fd58-41af-b6a2-41b53ef14373.png)
While their textmate scope are different, they share exactly the same token type and cannot be configured separately (the one in function definition has an extra modifier `parameter.declaration` so I can handle it separately)

**Comment by ZeroRin** [comment]
Now I got a dirty solution: I removed "variable" from the theme by directly modifying `dark_plus.json`:
```json
{
	"name": "Variable and parameter name",
	"scope": [
		// "variable",
		"meta.definition.variable.name",
		"support.variable",
		"entity.name.variable",
		"constant.other.placeholder", // placeholders in strings
	],
	"settings": {
		"foreground": "#9CDCFE"
	}
},
```
as the style `parameter` by default uses `variable.parameter`, once this is no longer available, semantic token type loses its effect and textmate rules are finally respected.
However, the solution is dirty because I directly modified the theme files instead of writing settings (I have no idea how to remove existing rules using settings) and because I need to add rule for every kind of `parameter` according to their textmate.

**Reply by rchiodo** [reply]
That does seem like the appropriate thing to me to do. You really want your own custom theme. Using that should allow to control the colors a lot more.
